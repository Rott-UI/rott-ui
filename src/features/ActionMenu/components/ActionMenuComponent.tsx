// React Imports
import React, {FC} from 'react'

// Component Imports
import {Item, Pressable, Label, Button, List, ActionMenu} from '../..'
import {ActionMenuStyles} from '../styles'
import {ActionMenuProps, ActionModel} from '../models'

// Constant Imports

// Util and Lib Imports

/**
 * Action Menu
 *
 * @param title Başlık alanı
 * @param subTitle Alt başlık alanı
 * @param data ActionModel[] formatında veri listesi
 * @param visible ActionMenu komponentin bulunduğu yerdeki Show state i
 * @returns Action Menu Renderlanır
 *
 */
export const ActionMenuComponent: FC<ActionMenuProps> = ({
  title,
  subTitle,
  data,
  itemHeight,
  maxItem,
  separatorTotalHeight,
}) => {
  const itemCount = (data?.length ?? 0) > maxItem! ? maxItem! : data?.length ?? 0

  // Listede renderlanacak Item
  const renderItem = ({
    title: renderItemTitle,
    name: renderItemName,
    action: onAction,
  }: ActionModel) => (
    <Pressable
      testID={`action-menu-data-${renderItemName}-test-id`}
      height={itemHeight}
      justifyContentCenter
      alignItemsCenter
      onPress={() => onAction()}>
      <Label fontSize='lg' variant='grey-900' fontWeight={500}>
        {renderItemTitle}
      </Label>
    </Pressable>
  )

  return (
    <Item size='full' style={ActionMenuStyles().actionMenuContainer} alignItemsCenter>
      <Item
        width={342}
        backgroundColor={COLOURS.GREY100}
        borderTopStartRadius={!(title || subTitle) ? 12 : 0}
        borderTopEndRadius={!(title || subTitle) ? 12 : 0}
        borderBottomStartRadius={12}
        borderBottomEndRadius={12}
        marginBottom={16}>
        <List
          testID='action-menu-test-id'
          style={{minHeight: 2}}
          width={342}
          height={itemHeight! * itemCount + separatorTotalHeight!}
          data={data}
          renderItem={({item}) => renderItem(item)}
          horizontal={false}
          separatorVariant='neutral-alpha-200'
          renderSeparator
          headerSeparator
          footerSeparator
          estimatedItemSize={display.normalize(itemHeight!, 'height')}
          showsVerticalScrollIndicator={false}
          scrollEnabled={(data?.length ?? 0) > maxItem!}
        />
      </Item>

      <Button
        variant='primary'
        onPress={() => ActionMenu.hideActionMenu()}
        fontSize='xl'
        size='full'
        testID='action-menu-cancel-test-id'
        marginBottom={21}>
        {translator('COMMON.CANCEL')}
      </Button>
    </Item>
  )
}
