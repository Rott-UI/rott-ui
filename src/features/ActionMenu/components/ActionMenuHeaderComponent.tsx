// React Imports
import React, {FC} from 'react'

// Util and Lib Imports
import {Item, Label} from '../..'
import {themeConfig} from '../../../providers'

interface ActionMenuHeaderProps {
  title?: Nullable<string>
  subTitle?: Nullable<string>
}

export const ActionMenuHeaderComponent: FC<ActionMenuHeaderProps> = ({title, subTitle}) => {
  return (
    <Item size='full' alignItemsCenter>
      <Item testID='action-menu-header-test-id' width={342}>
        {(title || subTitle) && (
          <Item
            justifyContentCenter
            alignItemsCenter
            size='full'
            backgroundColor={themeConfig.colors['grey-100']}
            borderTopStartRadius={12}
            borderTopEndRadius={12}
            height={54}>
            {title && (
              <Label
                fontSize='xl'
                variant='grey-900'
                fontWeight={700}
                marginBottom={subTitle ? 4 : undefined}>
                {title}
              </Label>
            )}
            {subTitle && (
              <Label fontSize='sm' variant='grey-900' fontWeight={500}>
                {subTitle}
              </Label>
            )}
          </Item>
        )}
      </Item>
    </Item>
  )
}
