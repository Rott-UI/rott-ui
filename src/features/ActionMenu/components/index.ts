export * from './ActionMenuComponent'
export * from './ActionMenuHeaderComponent'
