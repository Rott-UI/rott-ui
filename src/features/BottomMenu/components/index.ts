export * from './BottomMenu'
export * from './BottomMenuItem'
