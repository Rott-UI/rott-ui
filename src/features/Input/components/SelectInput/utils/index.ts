export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
