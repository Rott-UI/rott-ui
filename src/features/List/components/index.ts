export * from './List'
export * from './FavoriteSwipeComponent'
