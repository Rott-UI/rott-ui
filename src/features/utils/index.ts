export * from './commonUiTestExtension'
export * from './commonUiStyleProperties'
