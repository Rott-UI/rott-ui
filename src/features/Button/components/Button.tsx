// React Imports
import {FC, PropsWithChildren} from 'react'

// React Native Imports
import {TouchableOpacity, TouchableOpacityProps, StyleSheet, ActivityIndicator} from 'react-native'

// Component Imports
import {CommonUiProps, Size} from '../../models'
import {Label} from '../../Label'
import {Icon, IconTypes} from '../../Icon'

// Constant Imports
import {COLOURS} from '../../../constants'

// Model Imports
import {ButtonIconProps, ButtonImageProps} from '../models'

// Style Imports
import {ButtonStyles} from '../styles/Button.style'

// Util and Lib Imports
import {buttonSizeNormalizer} from '../../../utils/buttonSizeNormalizer'
import {display} from '../../../utils'
import {Image} from '../../Image'

interface ButtonProps
  extends TouchableOpacityProps,
    Omit<CommonUiProps, 'size'>,
    PropsWithChildren {
  color?: string
  isLoading?: boolean
  loadingText?: string
  backgroundColor?: string
  size?:
    | Size
    | {
        width?: Omit<Size, 'xl' | 'xxl'> | number
        height?: Omit<Size, 'xl' | 'xxl'> | number
      }

  leftImage?: ButtonImageProps
  rightImage?: ButtonImageProps
  leftIcon?: ButtonIconProps
  rightIcon?: ButtonIconProps

  height?: Size
  width?: Size
  borderWidth?: number
  borderColor?: string

  circle?: boolean
}

export const Button: FC<ButtonProps> = ({
  variant = 'primary',
  color,
  size = {height: 'lg'},
  isLoading,
  loadingText,
  flex = (typeof size === 'object' && size?.width) || typeof size === 'string' ? 0 : 1,

  borderWidth,
  borderColor,
  flexGrow,
  circle = false,

  leftImage,
  rightImage,

  leftIcon,
  rightIcon,

  height = 'full',
  width = 'full',

  children,
  style,
  ...props
}) => {
  return (
    <TouchableOpacity
      activeOpacity={0.8}
      style={StyleSheet.flatten([
        ButtonStyles({
          variant,
          color,
          size,
          flex,
          flexGrow,
          borderWidth,
          borderColor,
          circle,
          width,
          height,
          ...props,
        }).defaultButtonStyle,
        style,
      ])}
      {...props}>
      {leftIcon && !isLoading && (
        <Icon
          {...leftIcon}
          absolute={leftIcon.absolute}
          left={leftIcon.absolute ? display.normalize(24) : undefined}
          testID='button-left-icon-test-id'
          marginRight={leftIcon.absolute ? 0 : 8}
          variant={leftIcon?.variant}
          name={leftIcon?.name as IconTypes}
          width={
            leftIcon?.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            leftIcon?.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
        />
      )}

      {leftImage && !isLoading && (
        <Image
          testID='button-left-image-test-id'
          left={leftImage.absolute ? display.normalize(24) : undefined}
          name={leftImage.name}
          absolute={leftImage.absolute}
          marginRight={leftImage.absolute ? 0 : 8}
          width={
            leftImage.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            leftImage.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          tintColor={leftImage.tintColor}
          resizeMode={leftImage.resizeMode}
        />
      )}

      {!isLoading && (
        <Label
          numberOfLines={1}
          textCenter
          fontFamily='Markpro'
          fontSize={
            typeof size === 'object'
              ? buttonSizeNormalizer(size.height as Size).fontSize
              : buttonSizeNormalizer(size).fontSize
          }
          style={ButtonStyles({variant}).buttonTextStyle}>
          {children}
        </Label>
      )}

      {isLoading && (
        <ActivityIndicator color={COLOURS.WHITE} testID='button-loading-indicator-test-id' />
      )}

      {isLoading && loadingText && (
        <Label testID='button-loading-text-test-id' style={ButtonStyles({variant}).buttonTextStyle}>
          {loadingText}
        </Label>
      )}

      {rightIcon && !isLoading && (
        <Icon
          {...rightIcon}
          absolute={rightIcon.absolute}
          right={rightIcon.absolute ? display.normalize(24) : undefined}
          testID='button-right-icon-test-id'
          marginLeft={rightIcon.absolute ? 0 : 8}
          variant={rightIcon?.variant}
          name={rightIcon?.name as IconTypes}
          width={
            rightIcon.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            rightIcon.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
        />
      )}

      {rightImage && !isLoading && (
        <Image
          testID='button-right-image-test-id'
          right={rightImage.absolute ? display.normalize(24) : undefined}
          name={rightImage.name}
          absolute={rightImage.absolute}
          marginRight={rightImage.absolute ? 0 : 8}
          width={
            rightImage.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            rightImage.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          tintColor={rightImage.tintColor}
          resizeMode={rightImage.resizeMode}
        />
      )}
    </TouchableOpacity>
  )
}
