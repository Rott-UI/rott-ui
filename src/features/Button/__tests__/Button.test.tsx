// Component Imports
import {Button} from '../components'

// Util and Lib Imports
import {colorFromVariant, fireEvent, render} from '../../../utils'
import {commonUiTestExtension} from '../../utils/commonUiTestExtension'

const testId = {
  buttonTestId: 'button-test-id',
  buttonLoadingIndicatiorTestId: 'button-loading-indicator-test-id',
  buttonLoadingTextTestId: 'button-loading-text-test-id',
  buttonLeftIconTestId: 'button-left-icon-test-id',
  buttonLeftImageTestId: 'button-left-image-test-id',
  buttonRightIconTestId: 'button-right-icon-test-id',
  buttonRightImageTestId: 'button-right-image-test-id',
  buttonText: 'Test Button',
}

describe('Button -> Custom Component', () => {
  it('butonun snapshotı ile eşleşmeli', () => {
    const {buttonTestId, buttonText} = testId
    const rendered = render(<Button testID={buttonTestId}>{buttonText}</Button>)

    expect(rendered).toMatchSnapshot()
  })

  it('buton ekranda olmalı ancak loading iconu görünmemeli', () => {
    const {buttonTestId, buttonText, buttonLoadingIndicatiorTestId} = testId
    const {getByTestId, queryByTestId} = render(<Button testID={buttonTestId}>{buttonText}</Button>)

    const buttonElement = getByTestId(buttonTestId)
    const loadingIndicator = queryByTestId(buttonLoadingIndicatiorTestId)

    expect(buttonElement).toBeOnTheScreen()
    expect(loadingIndicator).not.toBeOnTheScreen()
  })

  it('loading text ve loading icon ile buton render olmalı', () => {
    const {buttonTestId, buttonText, buttonLoadingTextTestId, buttonLoadingIndicatiorTestId} =
      testId
    const renderedButton = render(<Button testID={buttonTestId}>{buttonText}</Button>)

    const {getByTestId, rerender} = renderedButton
    rerender(
      <Button testID={buttonTestId} isLoading loadingText='Loading'>
        {buttonText}
      </Button>
    )
    const buttonElement = getByTestId(buttonTestId)
    const loadingIndicator = getByTestId(buttonLoadingIndicatiorTestId)
    const loadingText = getByTestId(buttonLoadingTextTestId)

    expect(buttonElement).toBeOnTheScreen()
    expect(loadingIndicator).toBeOnTheScreen()
    expect(loadingText).toBeOnTheScreen()
  })

  it('buton disableken tıklanamamlı', () => {
    const {buttonTestId, buttonText} = testId
    const onPressMock = jest.fn()
    const {getByTestId} = render(
      <Button testID={buttonTestId} disabled onPress={onPressMock}>
        {buttonText}
      </Button>
    )

    const buttonElemet = getByTestId(buttonTestId)
    fireEvent.press(buttonElemet)

    expect(onPressMock).not.toBeCalled()
  })

  it('left icon propertysi verildiğinde ekranda doğru icon ile renderlanmalı', () => {
    const {buttonTestId, buttonText, buttonLeftIconTestId} = testId
    const {getByTestId} = render(
      <Button testID={buttonTestId} leftIcon={{name: 'REMOVE_CIRCLE'}}>
        {buttonText}
      </Button>
    )

    const leftIconElement = getByTestId(buttonLeftIconTestId)

    expect(leftIconElement).toBeOnTheScreen()
  })

  it('left image propertysi verildiğinde ekranda doğru image ile renderlanmalı', () => {
    const {buttonTestId, buttonText, buttonLeftImageTestId} = testId
    const {getByTestId} = render(
      <Button testID={buttonTestId} leftImage={{name: 'LEFT_ARROW_ICON', absolute: true}}>
        {buttonText}
      </Button>
    )

    const leftIconElement = getByTestId(buttonLeftImageTestId)

    expect(leftIconElement).toBeOnTheScreen()
    expect(leftIconElement).toHaveProp('source')
  })

  it('button loading durumundayken verilen left icon görünmemeli', () => {
    const {buttonText, buttonLeftIconTestId} = testId
    const {queryByTestId} = render(
      <Button leftIcon={{name: 'REMOVE_CIRCLE'}} isLoading>
        {buttonText}
      </Button>
    )

    const leftIcon = queryByTestId(buttonLeftIconTestId)

    expect(leftIcon).not.toBeOnTheScreen()
  })

  it('buton ekranda olmalı ve outline variantı ile renderlanmalı', () => {
    const {buttonTestId, buttonText} = testId
    const {getByTestId} = render(
      <Button variant='primary-outline' testID={buttonTestId}>
        {buttonText}
      </Button>
    )

    const buttonElement = getByTestId(buttonTestId)

    expect(buttonElement).toBeOnTheScreen()
    expect(buttonElement).toHaveStyle({
      backgroundColor: 'transparent',
      borderColor: colorFromVariant('primary'),
    })
  })

  it('right icon propertysi verildiğinde ekranda doğru icon ile renderlanmalı', () => {
    const {buttonTestId, buttonText, buttonRightIconTestId} = testId
    const {getByTestId} = render(
      <Button testID={buttonTestId} rightIcon={{name: 'REMOVE_CIRCLE'}}>
        {buttonText}
      </Button>
    )

    const rightIconElement = getByTestId(buttonRightIconTestId)

    expect(rightIconElement).toBeOnTheScreen()
  })

  it('right image propertysi verildiğinde ekranda doğru image ile renderlanmalı', () => {
    const {buttonTestId, buttonText, buttonRightImageTestId} = testId
    const {getByTestId} = render(
      <Button testID={buttonTestId} rightImage={{name: 'RIGHT_ARROW_ICON', absolute: true}}>
        {buttonText}
      </Button>
    )

    const rightImageElement = getByTestId(buttonRightImageTestId)

    expect(rightImageElement).toBeOnTheScreen()
    expect(rightImageElement).toHaveProp('source')
  })

  it('button loading durumundayken verilen right icon görünmemeli', () => {
    const {buttonText, buttonRightIconTestId} = testId
    const {queryByTestId} = render(
      <Button rightIcon={{name: 'REMOVE_CIRCLE'}} isLoading>
        {buttonText}
      </Button>
    )

    const rightIcon = queryByTestId(buttonRightIconTestId)

    expect(rightIcon).not.toBeOnTheScreen()
  })
})

commonUiTestExtension(<Button testID={testId.buttonTestId} />, testId.buttonTestId)
