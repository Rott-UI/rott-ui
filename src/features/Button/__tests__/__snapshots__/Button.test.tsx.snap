// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button -> Custom Component butonun snapshotı ile eşleşmeli 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(63, 182, 210, 1)",
        "borderColor": "white",
        "borderRadius": 8,
        "borderWidth": undefined,
        "bottom": undefined,
        "color": undefined,
        "flex": 1,
        "flexDirection": "row",
        "flexGrow": undefined,
        "gap": undefined,
        "height": 108,
        "justifyContent": "center",
        "left": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "maxHeight": undefined,
        "maxWidth": undefined,
        "opacity": 1,
        "overflow": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "paddingVertical": undefined,
        "position": "relative",
        "right": undefined,
        "top": undefined,
        "width": 658,
        "zIndex": undefined,
      }
    }
    testID="button-test-id"
  >
    <Text
      numberOfLines={1}
      style={
        {
          "backgroundColor": undefined,
          "bottom": undefined,
          "color": "rgba(255, 255, 255, 1)",
          "flexGrow": undefined,
          "flexShrink": 1,
          "fontFamily": "Markpro-Bold",
          "fontSize": 41,
          "fontWeight": undefined,
          "gap": undefined,
          "height": undefined,
          "left": undefined,
          "letterSpacing": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "maxHeight": undefined,
          "maxWidth": undefined,
          "overflow": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
          "position": undefined,
          "right": undefined,
          "textAlign": "center",
          "textAlignVertical": "center",
          "top": undefined,
          "width": undefined,
          "zIndex": undefined,
        }
      }
    >
      Test Button
    </Text>
  </View>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
]
`;
